core/coreapi/dag.go

```diff
@@ -56,6 +61,244 @@ func (api *dagAPI) Session(ctx context.Context) ipld.NodeGetter {
+	// it is simply a way to relieve pressure on the blockstore
+	// similar to pinner.Pin/pinner.Flush
+	batch := ipld.NewBatch(ctx, api)
+
+	roots := make(map[cid.Cid]struct{})
+
+	it := dir.Entries()
+	for it.Next() {
+
+		file := files.FileFromEntry(it)
+		if file == nil {
+			ret <- importResult{err: errors.New("expected a file handle")}
+			return
+		}
+
+		// wrap a defer-closer-scope
+		//
+		// every single file in it() is already open before we start
+		// just close here sooner rather than later for neatness
+		// and to surface potential errors writing on closed fifos
+		// this won't/can't help with not running out of handles
+		err := func() error {
+			defer file.Close()
+
+			car, err := gocar.NewCarReader(file)
+			if err != nil {
+				return err
+			}
+
+			// Be explicit here, until the spec is finished
+			if car.Header.Version != 1 {
+				return errors.New("only car files version 1 supported at present")
+			}
+
+			for _, c := range car.Header.Roots {
+				roots[c] = struct{}{}
+			}
+
+			for {
+				block, err := car.Next()
+				if err != nil && err != io.EOF {
+					return err
+				} else if block == nil {
+					break
+				}
+
+				// the double-decode is suboptimal, but we need it for batching
+				nd, err := ipld.Decode(block)
+				if err != nil {
+					return err
+				}
+
+				if err := batch.Add(ctx, nd); err != nil {
+					return err
+				}
+			}
+
+			return nil
+		}()
+
+		if err != nil {
+			ret <- importResult{err: err}
+			return
+		}
+	}
+
+	if err := it.Err(); err != nil {
+		ret <- importResult{err: err}
+		return
+	}
+
+	if err := batch.Commit(); err != nil {
+		ret <- importResult{err: err}
+		return
+	}
+
+	ret <- importResult{roots: roots}
+}
+
+// Export exports the DAG rooted at a given CID into a CAR file. The error channel will return at most one error and
+// the reader is usable before the error channel has been closed.
+func (api *dagAPI) Export(ctx context.Context, c cid.Cid) (io.ReadCloser, <-chan error) {
+	// Code disabled until descent-issue in go-ipld-prime is fixed
+	// https://github.com/ribasushi/gip-muddle-up
+	//
+	// sb := gipselectorbuilder.NewSelectorSpecBuilder(gipfree.NodeBuilder())
+	// car := gocar.NewSelectiveCar(
+	// 	req.Context,
+	// 	<needs to be fixed to take format.NodeGetter as well>,
+	// 	[]gocar.Dag{gocar.Dag{
+	// 		Root: c,
+	// 		Selector: sb.ExploreRecursive(
+	// 			gipselector.RecursionLimitNone(),
+	// 			sb.ExploreAll(sb.ExploreRecursiveEdge()),
+	// 		).Node(),
+	// 	}},
+	// )
+	// ...
+	// if err := car.Write(pipeW); err != nil {}
+
+	pipeR, pipeW := io.Pipe()
+
+	errCh := make(chan error, 1) // we only report the 1st error
+	go func() {
+		defer close(errCh)
+		var err error
+		if err = gocar.WriteCar(
+			ctx,
+			api.Session(ctx),
+			[]cid.Cid{c},
+			pipeW,
+		); err != nil {
+			errCh <- err
+		}
+
+		if closeErr := pipeW.Close(); closeErr != nil && err == nil {
+			errCh <- fmt.Errorf("stream flush failed: %s", closeErr)
+		}
```

Comment:
I kept the code mostly the same, except that we now only return a single error instead of returning two errors assuming that the caller will only process one of them which is what we did before. This required moving this pipe closure + error checking out of the defer statement.

We could also just wait on both functions to complete and return a multierror.